### PR TITLE
Fix implementation of exciter, new turbine governor model and refactor VBR models

### PR DESCRIPTION
**Pull request by martinmoraga**
- fixes implementation of exciter DC1A model 
- adds new turbine governor TGOV1 (Ref: Milano, Power system modelling and scripting p. 358)
- adds new validation notebook for controllers: [SP_Validation_ReducedOrderSG_VBR_SMIB_Fault_withControllers.ipynb](https://github.com/sogno-platform/dpsim/blob/SG_VBR_CurrentSource/examples/Notebooks/Circuits/SP_Validation_ReducedOrderSG_VBR_SMIB_Fault_withControllers.ipynb)
- partially solves points of #75 
- enables choice between thevenin and norton representation of VBR models, default is norton now
- adds reduced order SG models, exciter and turbine governor to dpsimpy

### CODE REVIEW

**Review by m-mirz — APPROVED** [review]

**Comment by JTS22** [comment]
@martinmoraga Can you maybe push the pybind-binding code that does not work for you, along with the error message? Normally, you should be able to just bind the abstract base class along with any child classes like this (assuming `stepInPerUnit` is changed to be public):
```cpp
	py::class_<CPS::DP::Ph1::ReducedOrderSynchronGeneratorVBR, std::shared_ptr<CPS::DP::Ph1::ReducedOrderSynchronGeneratorVBR>, CPS::SimPowerComp<CPS::Complex>>(mDPPh1, "BaseSynchronGeneratorReducedOrder", py::multiple_inheritance())
		.def("set_modelling_approach", &CPS::DP::Ph1::ReducedOrderSynchronGeneratorVBR::setModellingApproach)
		.def("step_in_per_unit", &CPS::DP::Ph1::ReducedOrderSynchronGeneratorVBR::stepInPerUnit);

	py::class_<CPS::DP::Ph1::SynchronGenerator3OrderVBR, std::shared_ptr<CPS::DP::Ph1::SynchronGenerator3OrderVBR>, CPS::DP::Ph1::ReducedOrderSynchronGeneratorVBR>(mDPPh1, "SynchronGenerator3OrderVBR", py::multiple_inheritance())
		.def(py::init<std::string>());
```

The base class is then non-constructible in Python but because Python only handles the shared pointers, methods can still take and return instances of the base class.
For a more comprehensive example, you can check the bindings in `dpsim/src/pybind/Attributes.cpp` where both `AttributeBase` and `Attribute<T>` are abstract classes that are bound to Python.

File: dpsim-models/include/dpsim-models/Base/Base_ReducedOrderSynchronGenerator.h [hunk 1]
```diff
@@ -58,6 +58,8 @@ namespace Base {
 			Real mEf_prev;
 
 		protected:
+			//
+			ModApproach mModApproach;
```
[the remark this answers is not in the document]
Comment (martinmoraga) [reply dpsim-models/include/dpsim-models/Base/Base_ReducedOrderSynchronGenerator.h]: done!

File: dpsim-models/include/dpsim-models/DP/DP_Ph1_ReducedOrderSynchronGeneratorVBR.h [hunk 2]
```diff
@@ -18,10 +18,14 @@ namespace Ph1 {
...
 		public Base::ReducedOrderSynchronGenerator<Complex>,
 		public MNAVariableCompInterface {
+
 	public:
         // Common elements of all VBR models
 		/// voltage behind reactance phase a
-        const Attribute<Complex>::Ptr mEvbr;
+        Complex mEvbr;
+		/// norton equivalent current of mEvbr
+		Matrix mIvbr;
```
[the remark this answers is not in the document]
Comment (martinmoraga) [reply dpsim-models/include/dpsim-models/DP/DP_Ph1_ReducedOrderSynchronGeneratorVBR.h]: done!

File: dpsim-models/include/dpsim-models/SP/SP_Ph1_SynchronGenerator4OrderDCIM.h [hunk 3]
```diff
@@ -44,25 +46,25 @@ namespace Ph1 {
...
 	public:
 		///
-		SynchronGenerator4OrderDCIM(String uid, String name, Logger::Level logLevel = Logger::Level::off);
-		///
-		SynchronGenerator4OrderDCIM(String name, Logger::Level logLevel = Logger::Level::off);
+		SynchronGenerator4OrderDCIM(const String & uid, const String & name, Logger::Level logLevel = Logger::Level::off);
 		///
-		SimPowerComp<Complex>::Ptr clone(String name);
+		SynchronGenerator4OrderDCIM(const String & name, Logger::Level logLevel = Logger::Level::off);
+		 ///
```
[the remark this answers is not in the document]
Comment (martinmoraga) [reply dpsim-models/include/dpsim-models/SP/SP_Ph1_SynchronGenerator4OrderDCIM.h]: done!

File: dpsim-models/include/dpsim-models/SP/SP_Ph1_ReducedOrderSynchronGeneratorVBR.h [hunk 4]
```diff
@@ -20,48 +20,55 @@ namespace Ph1 {
 		public MNAVariableCompInterface {
 	public:
         // Common elements of all VBR models
-        /// voltage behind reactance phase a
-        const Attribute<Complex>::Ptr Evbr;
-    protected:
+		/// voltage behind reactance phase a
+        Complex mEvbr;
+		/// norton equivalent current of mEvbr
+		Matrix mIvbr;
```
[the remark this answers is not in the document]
Comment (martinmoraga) [reply dpsim-models/include/dpsim-models/SP/SP_Ph1_ReducedOrderSynchronGeneratorVBR.h]: done!

File: dpsim-models/src/Base/Base_ReducedOrderSynchronGenerator.cpp [hunk 5]
```diff
@@ -125,18 +163,87 @@ void Base::ReducedOrderSynchronGenerator<VarType>::setOperationalParametersPerUn
...
+		mYd = 0;
+		mYq = 0;
+	}
+
+	Real Zq_t = mLd - mLd_t - mYd;
+	Real Zd_t = mLq - mLq_t - mYq;
+	Real Zq_s = mLd_t - mLd_s + mYd;
+	Real Zd_s = mLq_t - mLq_s + mYq;
+
+	mAd_t = mTimeStep * Zd_t / (2 * mTq0_t + mTimeStep);
```
[the remark this answers is not in the document]
Comment (dinkelbachjan) [reply dpsim-models/src/Base/Base_ReducedOrderSynchronGenerator.cpp:216]: Added as follow-up to #75

**Review by dinkelbachjan — APPROVED** [review]

**Pull request merged**